### ISSUE TITLE
Big, big rename

### ISSUE THREAD

**Issue by tomjaguarpaw** [tomjaguarpaw/haskell-opaleye#326]
I would like to release a version of Opaleye with better names for a lot of types.  Initially this will be a non-breaking release with type synonyms.  I would like to gather input from interested parties about the names.  The following are useful for reference:

* https://github.com/tomjaguarpaw/haskell-opaleye/issues/229

* https://github.com/tomjaguarpaw/haskell-opaleye/issues/286

* https://github.com/tomjaguarpaw/haskell-opaleye/issues/299

Done | Old | New
-----|-----|-------
&nbsp; | Column | Field
✓ | Query | Select
✓ | QueryArr | SelectArr
✓ | PG*Type*  | Sql*Type*
✓ | Constant | ToFields
✓ | QueryRunner | FromFields
✓ | QueryRunnerColumn | FromField
✓ | QueryRunnerColumnDefault | DefaultFromField
✓ | Table | Table
✓ | TableProperties | TableFields

EDIT: Before settling on `Field` we considered *X* as follows:

Candidates for *X* include `Field`, `Column` and `Value`.  `Field` is attested by point 5 of the second list of https://dba.stackexchange.com/a/183980/28020, posgresql-simple and [libpq](https://www.postgresql.org/docs/current/static/libpq-exec.html).

Interested parties include, but are not limited to, @ocharles, @hesselink, @bergmark, @saurabhnanda, @k0001, @tstat

**Comment by ocharles** [comment]
`rel8` uses `Expr` for `Column` - is that out of the question?

Also, with a move to `SqlType`, are you planning on supporting more than PostgreSQL?

**Comment by tomjaguarpaw** [comment]
> `rel8` uses `Expr` for `Column` - is that out of the question?

I'd rather that consumers didn't know that it is an expression.  That's really just an implementation detail.  I'm open to further discussion on the issue if you think it's important.

> Also, with a move to `SqlType`, are you planning on supporting more than PostgreSQL?

There's a languishing SQLite branch which already uses `SqlType` (as far as I remember).  I'm not planning on doing much work to resurrect it and I'm not planning on supporting any other RDBMSes, but it is possible that it will happen in the future so I'd rather be prepared for that eventuality.

**Comment by saurabhnanda** [comment]
+100 for this rename. 

May I also share another perspective? Every time I'm showing Opaleye to someone I get this question and I'm unable to give a satisfactory answer -- if I have setup my table mapping, and I'm querying/updating/inserting table X, why do I have to give a type annotation for the return value? Why can't haskell/opaleye figure out the return type on its own? 

In the context of this thread, would this be some sort of fun-dep in the typeclass of DefaultFromX? 

Further, how does one really use FromX instead of DefaultFromX? Drawing a parallel from JSON encoding/decoding, the ToJSON and FromJSON instances *are* the default functions for encoding/decoding values of a particular type. If you don't need the default behaviour you can call encodeWith and decodeWith and give it a non-default parser. What's the parallel in the Opaleye world? Consequently, do we really need both, FromX and DefaultFromX typeclasses?

**Comment by tomjaguarpaw** [comment]
> Further, how does one really use FromX instead of DefaultFromX?

This question doesn't belong on this issue so I migrated it to https://github.com/tomjaguarpaw/haskell-opaleye/issues/327.

**Comment by tomjaguarpaw** [comment]
> if I have setup my table mapping, and I'm querying/updating/inserting table X, why do I have to give a type annotation for the return value?

This question doesn't belong here, and it has lots of different answers depending on what you mean by "why", but the very brief answer is "because Opaleye doesn't know if you want a `PGText` to be converted to a `String` or a `Text`".  If you want the longer answers please open a new issue.

**Comment by tomjaguarpaw** [comment]
I'd particularly like thoughts on `Column` vs `Field`.

**Comment by ocharles** [comment]
I can't say I like either of them. The literal "True" is not a column, and I don't even know what a field is. If you're opposed to `Expr` there is also `Attribute`, which has meaning in relational theory which *is* the framework we're working in, after all.

**Comment by tomjaguarpaw** [comment]
`Attribute` is the correct name for the whole column, according to [this Stack Overflow answer](https://dba.stackexchange.com/questions/183969/what-are-the-names-of-these-postgres-concepts/183980#183980).

**Comment by ocharles** [comment]
True, I would agree with that answer.

**Comment by ocharles** [comment]
Edit: Actually, scrap that. I don't like `Attribute`. "Adding" two int-valued attributes doesn't make any sense.

**Comment by tomjaguarpaw** [comment]
Why doesn't that make any sense?

**Comment by ocharles** [comment]
Doesn't make sense is maybe a bit strong, but I guess I view attributes as named things - but the whole concept of names doesn't exactly fit with Opaleye as we're passing references around. What would the "name" of the sum of two attributes be? It is a bit of a conflation of terms.

**Comment by tomjaguarpaw** [comment]
How about `Value`, as short for "attribute value".

**Comment by ocharles** [comment]
Value seems pretty reasonable!

**Comment by tomjaguarpaw** [comment]
Other libraries

* relational-record: No concrete type, but typeclasses `OperatorProjectable` and `ProjectableShowSql`
* beam: `QExpr`
* selda: `Col`
* haskelldb: `Expr`

**Comment by ocharles** [comment]
* rel8: `Expr` :smile:

**Comment by tomjaguarpaw** [comment]
Oh yeah!

**Comment by saurabhnanda** [comment]
`PGResult` from Postgres' C library calls them tuples & fields.

**Comment by tomjaguarpaw** [comment]
@saurabhnanda Can you point us to the docs for that?

**Comment by saurabhnanda** [comment]
https://www.postgresql.org/docs/current/static/libpq-exec.html - look at section 32.3.2

**Comment by tomjaguarpaw** [comment]
@saurabhnanda Very helpful.  Thanks!

**Comment by tomjaguarpaw** [comment]
Many good renamings are in [version 0.6.7003.1](https://hackage.haskell.org/package/opaleye-0.6.7003.1).  More exist in master and a couple more exist in branch [field](https://github.com/tomjaguarpaw/haskell-opaleye/tree/field).  It would be good if interested parties could check that things still work for them.

**Comment by tomjaguarpaw** [comment]
And now on Hackage https://hackage.haskell.org/package/opaleye-0.6.7004.0

**Comment by noxecane** [comment]
I have been able to build without any issues. Though most of my code is "abstracted" away with [this](https://gist.github.com/noxecane/5dfebdbf64fc1167fbc91a9cfcfcc8e9)

**Comment by tomjaguarpaw** [comment]
@noxecane Great, thanks for the confirmation.

**Comment by tomjaguarpaw** [comment]
Dear Opaleye friends,

In an effort to make the API more approachable I am removing or deprecating many confusing names.  This means that code written against 0.6 may break in 0.7.  i would appreciate it if users could try the branch `0.7` on their codebases, firstly so that they are well-prepared and secondly so that any unexpected breakage can be reported and resolved prior to release.

Thanks very much!

**Comment by tomjaguarpaw** [comment]
Direct link to `0.7` branch: https://github.com/tomjaguarpaw/haskell-opaleye/tree/0.7

**Comment by tomjaguarpaw** [comment]
Right, I'm intending to pull the trigger on `Column` -> `Field` early in the new year. I would welcome it if interested parties could test the `column-to-field` branch against their codebases. It should only require changing `Column a` to `Field a`/`Field_ n a` and `Column (Nullable a)` to `FieldNullable a`.

(I already know [the changes due in rel8](https://github.com/circuithub/rel8/compare/master...tomjaguarpaw:column-to-field?expand=1) are rather small)

**Comment by tomjaguarpaw** [comment]
I'll do this by roughly the end of the month, unless anyone shouts.

**Comment by tomjaguarpaw** [comment]
Completed as of https://github.com/tomjaguarpaw/haskell-opaleye/tree/version_0.9.0.0

**Comment by nh2** [comment]
@tomjaguarpaw It seems this change breaks comparison of `Nullable` fields:

* Before: `(.==) ::` [Column](https://hackage.haskell.org/package/opaleye-0.7.6.1/docs/Opaleye-Internal-Column.html#t:Column) a -> [Column](https://hackage.haskell.org/package/opaleye-0.7.6.1/docs/Opaleye-Internal-Column.html#t:Column) a -> [Field](https://hackage.haskell.org/package/opaleye-0.7.6.1/docs/Opaleye-Field.html#t:Field) [SqlBool](https://hackage.haskell.org/package/opaleye-0.7.6.1/docs/Opaleye-Internal-PGTypesExternal.html#t:SqlBool)
* After: `(.==) ::` [Field](https://hackage.haskell.org/package/opaleye-0.10.0.0/docs/Opaleye-Field.html#t:Field) a -> [Field](https://hackage.haskell.org/package/opaleye-0.10.0.0/docs/Opaleye-Field.html#t:Field) a -> [Field](https://hackage.haskell.org/package/opaleye-0.10.0.0/docs/Opaleye-Field.html#t:Field) [SqlBool](https://hackage.haskell.org/package/opaleye-0.10.0.0/docs/Opaleye-SqlTypes.html#t:SqlBool)
  * `Field` is non-nullable

How does one compare nullable fields now?

The tutorials don't seem to cover this.

**Comment by nh2** [comment]
> How does one compare nullable fields now?

I suspect the correct transformation is to to use [`MaybeFields`](https://www.stackage.org/haddock/lts-21.11/opaleye-0.9.7.0/Opaleye-MaybeFields.html) and [`.===`](https://www.stackage.org/haddock/lts-21.11/opaleye-0.9.7.0/Opaleye-Operators.html#v:.-61--61--61-), for example to compare with a column `columnApiKey` we can use:

```diff
-columnApiKey row .== toNullable (pgUniqueId apiKeyConstant)
+nullableToMaybeFields (columnApiKey row) .=== justFields (pgUniqueId apiKeyConstant))
```

**Comment by tomjaguarpaw** [comment]
I would say it *fixes* the comparison of `Nullable` fields since `(.==) :: Column a -> Column a -> Field SqlBool` was not type safe! If either argument was `NULL` the result would be `Null`, but the type system would incorrectly think it was non-null.

Yes, converting  to `MaybeFields` using `nullableToMaybeFields` seems like a good approach. If that doesn't work satisfactorily for you then please [open a new issue](https://github.com/tomjaguarpaw/haskell-opaleye/issues/new) and we'll try to find a better solution.
